File: extractors/typescript/function-catalog.mjs
#!/usr/bin/env node
// function-catalog.mjs
//
// Walks every .ts / .tsx source file under --root (and optionally --shared)
// and emits one JSON record per declared function-like construct:
// FunctionDeclaration (including overload heads), MethodDeclaration,
// ArrowFunction / FunctionExpression assigned to a named binding.
//
// Each record carries:
//   - body_hash / body_lines: duplication clustering (NULL when normalized
//     body is shorter than --min-body-lines; the row is still emitted so
//     signature-level queries see one-liners)
//   - params[].type_ref + return_ref: cross-catalog type-resolution joins
//     (#133 public-api-leaks.jq, future graph-view consumers)
//   - signature_index: per (file, name) overload-head discriminator
//   - is_test / touched_in_window / synthetic: universal flags matching
//     type-catalog
//
// See ../../docs/pipeline-contract.md > Function catalog for the v1.1 schema.

import ts from 'typescript';
import { createHash } from 'node:crypto';
import { readFileSync, writeFileSync, readdirSync } from 'node:fs';
import { execSync } from 'node:child_process';
import { join, relative, resolve, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { parseArgs } from 'node:util';
import {
  BUILTIN_TYPE_DENYLIST,
  extractReferences,
  genericsList,
} from './_lib/references.mjs';
import { isTestPath } from './_lib/paths.mjs';
import { EXT_RE, SKIP_DIRS } from './_lib/walk-predicate.mjs';

const EXTRACTOR_DIR = dirname(fileURLToPath(import.meta.url));
const EXTRACTOR_VERSION = JSON.parse(readFileSync(join(EXTRACTOR_DIR, 'package.json'), 'utf8')).version;
const SCHEMA_VERSION = '1.2';
const FINGERPRINT_V = 'shape_sig:1';

// v1.2 identity / provenance — see docs/pipeline-contract.md "Identity and
// provenance". See extractors/typescript/type-catalog.mjs::computeSourceSha
// for the long-form rationale; the bounded-walk + timeout pattern is shared.
function computeSourceSha() {
  try {
    const parent = dirname(EXTRACTOR_DIR);
    const env = {
      ...process.env,
      GIT_CEILING_DIRECTORIES: parent,
      GIT_TERMINAL_PROMPT: '0',
    };
    const opts = {
      cwd: EXTRACTOR_DIR,
      stdio: ['ignore', 'pipe', 'ignore'],
      encoding: 'utf8',
      timeout: 2000,
      env,
    };
    const sha = execSync('git rev-parse HEAD', opts).trim();
    if (!/^[0-9a-f]{40}$/.test(sha)) return unknownSha();
    const toplevel = execSync('git rev-parse --show-toplevel', opts).trim();
    if (!toplevel || !EXTRACTOR_DIR.startsWith(toplevel)) return unknownSha();
    return sha;
  } catch {
    return unknownSha();
  }
}
function unknownSha() {
  process.stderr.write('warning: extractor source not in a git checkout; source_sha recorded as "unknown"\n');
  return 'unknown';
}
const SOURCE_SHA = computeSourceSha();
const GENERATED_AT = new Date().toISOString();

// NUL-byte joined; see extractors/typescript/type-catalog.mjs::computeSymbolId
// for the rationale (slash-joined formula was not injective when package or
// file contained `/`, which they legitimately do).
function computeSymbolId(pkg, file, name, kind) {
  return createHash('sha1').update(`${pkg}\x00${file}\x00${name}\x00${kind}`).digest('hex');
}

const { values } = parseArgs({
  options: {
    root:             { type: 'string' },
    shared:           { type: 'string' },
    touched:          { type: 'string' },
    output:           { type: 'string' },
    'min-body-lines': { type: 'string', default: '3' },
    help:             { type: 'boolean', default: false },
  },
});

if (values.help || !values.root) {
  process.stderr.write(`usage: function-catalog.mjs --root <path> [--shared <path>] [--touched <json>] [--output <path>] [--min-body-lines <n>]

  --root             Required. Root of the codebase to scan.
  --shared           Optional. A secondary package root. Tagged as package="shared".
  --touched          Optional. JSON array of file paths (relative to --root) to mark touched_in_window=true.
  --output           Optional. Write JSON to this path. Default: stdout.
  --min-body-lines   Optional. Threshold below which body fields (body_hash,
                     body_lines, body_line_count, body_length) are emitted as
                     null (default 3). The row itself is still emitted —
                     signature-level queries need exported one-liner functions
                     to be visible.

Output: {"schema_version": "${SCHEMA_VERSION}", "extractor": {...}, "entries": [...]}.
Test files are always extracted; every row carries an \`is_test\` flag derived
from the file path. To exclude tests post-hoc, pipe through:
  jq '.entries | map(select(.is_test | not))'
`);
  process.exit(values.help ? 0 : 1);
}

const ROOT = resolve(values.root);
const SHARED = values.shared ? resolve(values.shared) : null;
const MIN_BODY_LINES = Number(values['min-body-lines']) || 3;
const TOUCHED = values.touched
  ? new Set(JSON.parse(readFileSync(values.touched, 'utf8')))
  : new Set();

// SKIP_DIRS and EXT_RE imported from _lib/walk-predicate.mjs so this walker
// and type-catalog.mjs stay in lock-step on what counts as a TS source file.

function walkDir(root) {
  const out = [];
  function walk(dir) {
    let entries;
    try { entries = readdirSync(dir, { withFileTypes: true }); }
    catch { return; }
    for (const e of entries) {
      const full = join(dir, e.name);
      if (e.isDirectory()) {
        if (e.name.startsWith('.')) continue;
        if (SKIP_DIRS.has(e.name)) continue;
        walk(full);
      } else if (e.isFile()) {
        if (!EXT_RE.test(e.name)) continue;
        out.push(full);
      }
    }
  }
  walk(root);
  return out;
}

// --- Body normalization (unchanged from v1.0) ---

function normalizeBody(text) {
  let stripped = text.replace(/\/\*[\s\S]*?\*\//g, '');
  stripped = stripped.replace(/\/\/[^\n]*/g, '');
  const lines = stripped
    .split('\n')
    .map((l) => l.replace(/\s+/g, ' ').trim())
    .filter((l) => l.length > 0);
  return lines;
}

function bodyHashOf(normLines) {
  return createHash('sha256').update(normLines.join('\n')).digest('hex');
}

function bodyTextOf(node, sf) {
  return node ? node.getText(sf) : '';
}

// --- Modifiers ---

function exportedMod(node) {
  return !!node.modifiers?.some((m) => m.kind === ts.SyntaxKind.ExportKeyword);
}

function asyncMod(node) {
  return !!node.modifiers?.some((m) => m.kind === ts.SyntaxKind.AsyncKeyword);
}

// --- Signature extraction ---
//
// `type_ref` is the simple-case sugar (bare TypeReferenceNode whose typeName
// is an unbound, non-built-in Identifier with no type arguments); derived
// from `type_refs` to avoid a second AST walk.

function singleRefSugar(typeNode, refs) {
  if (refs.length !== 1) return null;
  if (!typeNode || !ts.isTypeReferenceNode(typeNode)) return null;
  if (!ts.isIdentifier(typeNode.typeName)) return null;
  if (typeNode.typeArguments) return null;
  return refs[0].name;
}

function refsOfTypeNode(typeNode, sf, scope) {
  if (!typeNode) return [];
  const sink = new Set();
  extractReferences(typeNode, sf, scope, sink);
  return [...sink].sort().map((name) => ({ name, kind: 'type-ref' }));
}

function signaturesOf(node, sf) {
  const scope = new Set();
  const allRefs = new Set();
  if (node.typeParameters) {
    for (const tp of node.typeParameters) scope.add(tp.name.text);
    for (const tp of node.typeParameters) {
      if (tp.constraint) extractReferences(tp.constraint, sf, scope, allRefs);
    }
  }
  const params = (node.parameters ?? []).map((p) => {
    const type_refs = refsOfTypeNode(p.type, sf, scope);
    for (const ref of type_refs) allRefs.add(ref.name);
    return {
      name: p.name.getText(sf),
      type_ref: singleRefSugar(p.type, type_refs),
      type_refs,
    };
  });
  let return_ref = null;
  if (node.type) {
    const returnRefs = refsOfTypeNode(node.type, sf, scope);
    for (const ref of returnRefs) allRefs.add(ref.name);
    return_ref = singleRefSugar(node.type, returnRefs);
  }
  const references = [...allRefs].sort().map((name) => ({ name, kind: 'type-ref' }));
  return { params, return_ref, references };
}

// Emit body-derived fields together — all four gate on `longEnough` and must
// move in lockstep (silent asymmetry on body_lines would land here otherwise).
function bodyFields(normLines, longEnough) {
  if (!longEnough) {
    return { body_line_count: null, body_length: null, body_hash: null, body_lines: null };
  }
  return {
    body_line_count: normLines.length,
    body_length: normLines.join('\n').length,
    body_hash: bodyHashOf(normLines),
    body_lines: [...new Set(normLines)].sort(),
  };
}

// --- Per-file extraction ---

function extractFromFile(filePath, pkgName, pkgRoot) {
  const text = readFileSync(filePath, 'utf8');
  const isTsx = filePath.endsWith('.tsx');
  const sf = ts.createSourceFile(
    filePath,
    text,
    ts.ScriptTarget.Latest,
    true,
    isTsx ? ts.ScriptKind.TSX : ts.ScriptKind.TS,
  );
  const relPath = relative(pkgRoot, filePath);
  const isGenerated = /(^|\/)generated\//.test(relPath) || relPath.endsWith('.d.ts');
  const isTest = isTestPath(relPath);
  const touched = pkgName === 'main' && TOUCHED.has(relPath);

  // Per-file Map<name, next index>. Two FunctionDeclarations sharing a name
  // in the same file -- whether overload heads at the top level or nested
  // same-name functions -- get indices 0, 1, 2, ... in source order. The
  // scope is intentionally per-(file, name) rather than per-(scope, name) so
  // overload-aware queries can dedupe with a `(name, package, file)` key.
  const sigIndex = new Map();

  const results = [];

  function pushFunction({ node, name, kind, exported, isAsync, body }) {
    const normLines = body ? normalizeBody(bodyTextOf(body, sf)) : [];
    const longEnough = normLines.length >= MIN_BODY_LINES;
    const { line } = sf.getLineAndCharacterOfPosition(node.getStart(sf));
    const sig = signaturesOf(node, sf);
    const idx = sigIndex.get(name) ?? 0;
    sigIndex.set(name, idx + 1);
    results.push({
      package: pkgName,
      file: relPath,
      line: line + 1,
      generated: isGenerated,
      is_test: isTest,
      touched_in_window: touched,
      synthetic: false,
      name,
      kind,
      exported,
      async: isAsync,
      param_count: sig.params.length,
      param_names: sig.params.map((p) => p.name),
      ...bodyFields(normLines, longEnough),
      generics: genericsList(node) ?? '',
      params: sig.params,
      return_ref: sig.return_ref,
      references: sig.references,
      references_count: sig.references.length,
      signature_index: idx,
    });
  }

  function visit(node) {
    // Top-level FunctionDeclaration. body is null for overload heads -- still emit.
    if (ts.isFunctionDeclaration(node) && node.name) {
      pushFunction({
        node,
        name: node.name.text,
        kind: 'function',
        exported: exportedMod(node),
        isAsync: asyncMod(node),
        body: node.body ?? null,
      });
    }

    // class Foo { bar(...) { ... } }
    if (ts.isMethodDeclaration(node) && node.name && node.body) {
      const className = (() => {
        let p = node.parent;
        while (p && !ts.isClassLike(p)) p = p.parent;
        return p && p.name ? p.name.text : null;
      })();
      const methodName = node.name.getText(sf);
      const qualifiedName = className ? `${className}.${methodName}` : methodName;
      pushFunction({
        node,
        name: qualifiedName,
        kind: 'method',
        exported: !!className && exportedMod(node.parent),
        isAsync: asyncMod(node),
        body: node.body,
      });
    }

    // const foo = (args) => { ... }  OR  const foo = function(args) { ... }
    if (ts.isVariableStatement(node)) {
      const exported = exportedMod(node);
      for (const decl of node.declarationList.declarations) {
        if (!decl.initializer || !decl.name || !ts.isIdentifier(decl.name)) continue;
        let init = decl.initializer;
        while (ts.isAsExpression(init) || ts.isTypeAssertionExpression?.(init) || ts.isParenthesizedExpression(init)) {
          init = init.expression;
        }
        if (ts.isArrowFunction(init) && init.body) {
          pushFunction({
            node: init,
            name: decl.name.text,
            kind: 'arrow-function',
            exported,
            isAsync: asyncMod(init),
            body: init.body,
          });
        } else if (ts.isFunctionExpression(init) && init.body) {
          pushFunction({
            node: init,
            name: decl.name.text,
            kind: 'function-expression',
            exported,
            isAsync: asyncMod(init),
            body: init.body,
          });
        }
      }
    }

    ts.forEachChild(node, visit);
  }

  visit(sf);
  return results;
}

// --- Run ---

const mainFiles = walkDir(ROOT);
const sharedFiles = SHARED ? walkDir(SHARED) : [];

process.stderr.write(`main: ${mainFiles.length} files\n`);
if (SHARED) process.stderr.write(`shared: ${sharedFiles.length} files\n`);

const all = [];
let errors = 0;
for (const f of mainFiles) {
  try { all.push(...extractFromFile(f, 'main', ROOT)); }
  catch (e) { errors++; process.stderr.write(`  ERR ${f}: ${e.message}\n`); }
}
for (const f of sharedFiles) {
  try { all.push(...extractFromFile(f, 'shared', SHARED)); }
  catch (e) { errors++; process.stderr.write(`  ERR ${f}: ${e.message}\n`); }
}

process.stderr.write(`\nTotal functions: ${all.length} (errors: ${errors})\n`);
const byKind = {};
const byPkg = {};
let isTestCount = 0;
for (const e of all) {
  byKind[e.kind] = (byKind[e.kind] || 0) + 1;
  byPkg[e.package] = (byPkg[e.package] || 0) + 1;
  if (e.is_test) isTestCount++;
}
process.stderr.write('By kind:\n');
for (const [k, v] of Object.entries(byKind).sort((a, b) => b[1] - a[1])) {
  process.stderr.write(`  ${String(v).padStart(5)}  ${k}\n`);
}
process.stderr.write('By package:\n');
for (const [k, v] of Object.entries(byPkg)) {
  process.stderr.write(`  ${String(v).padStart(5)}  ${k}\n`);
}
process.stderr.write(`is_test entries: ${isTestCount}\n`);

// v1.2 symbol_id pass — sha1(package + "/" + file + "/" + name + "/" + kind), lowercase hex.
// For method records `name` is already qualified as `ClassName.methodName`.
for (const e of all) {
  e.symbol_id = computeSymbolId(e.package, e.file, e.name, e.kind);
}

const catalog = {
  schema_version: SCHEMA_VERSION,
  extractor: {
    language: 'typescript',
    name: 'function-catalog',
    version: EXTRACTOR_VERSION,
    source_sha: SOURCE_SHA,
  },
  fingerprint_v: FINGERPRINT_V,
  generated_at: GENERATED_AT,
  entries: all,
};

const json = JSON.stringify(catalog, null, 2);
if (values.output) {
  writeFileSync(values.output, json);
  process.stderr.write(`\nWrote ${values.output}\n`);
} else {
  process.stdout.write(json);
}

process.exit(mainFiles.length > 0 ? 0 : 1);

File: extractors/python/function_catalog.py
#!/usr/bin/env python3
"""function_catalog.py — Python AST extractor for function-level catalog data.

Walks every .py source file under --root (and optionally --shared) and emits
one JSON record per function / async function / method declaration. Each record
carries body_hash + body_lines for duplication clustering and a signature
projection (params, return_ref, references) for cross-catalog joins.

See ../../docs/pipeline-contract.md for the emitted schema.
"""

from __future__ import annotations

import argparse
import ast
import datetime
import hashlib
import json
import re
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))
from _lib import (  # noqa: E402
    EXTRACTOR_VERSION,
    FINGERPRINT_V,
    LANGUAGE,
    SCHEMA_VERSION,
    annotation_text,
    compute_source_sha,
    is_generated,
    is_test_path,
    make_extractor_block,
    sha256_hex,
    walk_python_files,
    write_catalog,
)
from type_catalog import BUILTIN_TYPE_DENYLIST, _collect_name_refs  # noqa: E402

DEFAULT_MIN_BODY_LINES = 3


def _symbol_id(package: str, file: str, name: str, kind: str, signature_index: int) -> str:
    """sha1 over (package, file, name, kind, signature_index) joined by NUL
    bytes. `signature_index` is included so `@typing.overload` heads (each a
    separate FunctionDef sharing one name + kind) get distinct ids."""
    h = hashlib.sha1()
    h.update(
        f"{package}\x00{file}\x00{name}\x00{kind}\x00{signature_index}".encode("utf-8")
    )
    return h.hexdigest()


def _normalize_body_lines(stmts: list[ast.stmt]) -> list[str]:
    """Render body statements as a deduplicated, sorted, whitespace-normalized
    list of source lines.

    Comments are stripped naturally because ast doesn't keep them. Docstrings
    (first-statement string Expr) are dropped because they're documentation,
    not behavior."""
    if not stmts:
        return []
    # Drop docstring if first statement is a string Expr.
    if (isinstance(stmts[0], ast.Expr)
            and isinstance(stmts[0].value, ast.Constant)
            and isinstance(stmts[0].value.value, str)):
        stmts = stmts[1:]
    rendered: list[str] = []
    for s in stmts:
        try:
            src = ast.unparse(s)
        except Exception:
            continue
        for raw_line in src.split("\n"):
            line = re.sub(r"\s+", " ", raw_line).strip()
            if line:
                rendered.append(line)
    # Dedupe + sort for Jaccard input (matches the TS extractor's convention).
    return sorted(set(rendered))


def _single_type_ref(ann: ast.AST | None) -> str | None:
    """Return the single-identifier form of a type annotation; None for
    primitives, unions, anonymous shapes."""
    if ann is None:
        return None
    if isinstance(ann, ast.Name):
        n = ann.id
        if n in BUILTIN_TYPE_DENYLIST:
            return None
        return n
    if isinstance(ann, ast.Attribute):
        n = ann.attr
        if n in BUILTIN_TYPE_DENYLIST:
            return None
        return n
    return None


def _decorator_simple_names(decorators: list[ast.expr]) -> set[str]:
    """Return the rightmost-identifier set of a function's decorator list. Matches
    the convention used by type_catalog._decorator_names but returns a set for
    membership checks; handles `@dec` and `@dec(arg)` and `@mod.dec` forms."""
    out: set[str] = set()
    for d in decorators:
        if isinstance(d, ast.Call):
            d = d.func
        if isinstance(d, ast.Name):
            out.add(d.id)
        elif isinstance(d, ast.Attribute):
            out.add(d.attr)
    return out


def _has_implicit_self(fn: ast.FunctionDef | ast.AsyncFunctionDef, inside_class: bool) -> bool:
    """True if the function's first positional arg is implicit (`self` / `cls`)
    in Python's calling convention. False for `@staticmethod` and for any
    function declared outside a class. `@classmethod` keeps the implicit first
    arg (it's `cls`)."""
    if not inside_class:
        return False
    decs = _decorator_simple_names(fn.decorator_list)
    return "staticmethod" not in decs


def _extract_params(
    fn: ast.FunctionDef | ast.AsyncFunctionDef,
    has_implicit_self: bool,
) -> tuple[list[dict], list[str]]:
    """Return (params_payload, param_names).

    When `has_implicit_self` is True, the FIRST positional arg (`self` / `cls`
    by convention) is dropped from both the payload and the names list so
    `param_count` matches the TS arity convention (caller-perspective)."""
    args = fn.args
    out: list[dict] = []
    names: list[str] = []

    # The full ordered arg list, in spec order: posonly, args, vararg, kwonly, kwarg.
    all_args: list[ast.arg] = []
    all_args.extend(args.posonlyargs)
    all_args.extend(args.args)
    if args.vararg:
        all_args.append(args.vararg)
    all_args.extend(args.kwonlyargs)
    if args.kwarg:
        all_args.append(args.kwarg)

    for i, a in enumerate(all_args):
        # Strip the implicit first arg (whatever the author named it) on
        # methods that have one; never strip on functions, staticmethods,
        # or after the first positional.
        if i == 0 and has_implicit_self:
            continue
        names.append(a.arg)
        type_ref = _single_type_ref(a.annotation)
        type_refs = _collect_name_refs(a.annotation) if a.annotation is not None else []
        out.append({
            "name": a.arg,
            "type_ref": type_ref,
            "type_refs": type_refs,
        })
    return out, names


def _walk_functions(tree: ast.Module):
    """Yield (qual_name, FunctionDef|AsyncFunctionDef, kind, inside_class) for
    every function declaration in the module. Class methods are qualified as
    Class.method. `inside_class` tells the caller whether the function's first
    positional arg is implicit-self by Python's calling convention (modulo
    `@staticmethod`)."""
    def walk(node: ast.AST, qual_prefix: str = "", inside_class: bool = False):
        for child in ast.iter_child_nodes(node):
            if isinstance(child, ast.ClassDef):
                inner_prefix = qual_prefix + child.name + "."
                yield from walk(child, qual_prefix=inner_prefix, inside_class=True)
            elif isinstance(child, (ast.FunctionDef, ast.AsyncFunctionDef)):
                qual = qual_prefix + child.name
                kind = "method" if inside_class else "function"
                yield qual, child, kind, inside_class
                # Walk into nested defs for closures; the prefix follows the dotted form.
                yield from walk(child, qual_prefix=qual + ".", inside_class=False)
    yield from walk(tree)


def extract_from_file(
    file_path: Path,
    pkg_name: str,
    pkg_root: Path,
    touched: set[str],
    min_body_lines: int,
) -> list[dict]:
    text = file_path.read_text(encoding="utf-8", errors="replace")
    try:
        tree = ast.parse(text, filename=str(file_path))
    except SyntaxError as exc:
        sys.stderr.write(f"  parse error {file_path}: {exc}\n")
        return []
    rel = str(file_path.relative_to(pkg_root))
    row_defaults = {
        "package": pkg_name,
        "file": rel,
        "language": LANGUAGE,
        "touched_in_window": pkg_name == "main" and rel in touched,
        "generated": is_generated(rel),
        "is_test": is_test_path(rel),
        "synthetic": False,
    }
    out: list[dict] = []
    # Per-(file, name) counter so two FunctionDefs sharing a name in the same
    # file -- @typing.overload heads above an implementation, or nested
    # same-name defs -- get signature_index 0, 1, 2, ... in source order.
    # Mirrors the TS function-catalog convention so (name, package, file)
    # dedupe + signature_index ordering both work uniformly.
    sig_index: dict[str, int] = {}
    for qual, fn, kind, inside_class in _walk_functions(tree):
        params, param_names = _extract_params(fn, _has_implicit_self(fn, inside_class))

        body_lines = _normalize_body_lines(fn.body)
        if len(body_lines) >= min_body_lines:
            body_text = "\n".join(body_lines)
            body_hash = sha256_hex(body_text)
            body_line_count = len(body_lines)
            body_length = len(body_text)
        else:
            body_hash = None
            body_lines = None  # type: ignore[assignment]
            body_line_count = None
            body_length = None

        return_ref = _single_type_ref(fn.returns)
        return_refs = _collect_name_refs(fn.returns) if fn.returns is not None else []

        # Function-level references: deduplicated union of all param type_refs +
        # the return-side ref set, minus generic names (Python has no easy way
        # to scope TypeVars, so we just rely on the denylist filtering).
        all_refs: dict[str, None] = {}
        for p in params:
            for r in p["type_refs"]:
                all_refs[r["name"]] = None
        for r in return_refs:
            all_refs[r["name"]] = None
        references = [{"name": k, "kind": "type-ref"} for k in sorted(all_refs.keys())]

        async_flag = isinstance(fn, ast.AsyncFunctionDef)
        exported = not qual.split(".")[-1].startswith("_")

        idx = sig_index.get(qual, 0)
        sig_index[qual] = idx + 1

        row = {
            **row_defaults,
            "name": qual,
            "kind": kind,
            "line": fn.lineno,
            "symbol_id": _symbol_id(pkg_name, rel, qual, kind, idx),
            "exported": exported,
            "async": async_flag,
            "param_count": len(params),
            "param_names": param_names,
            "body_hash": body_hash,
            "body_line_count": body_line_count,
            "body_length": body_length,
            "body_lines": body_lines,
            "generics": "",
            "params": params,
            "return_ref": return_ref,
            "references": references,
            "references_count": len(references),
            "signature_index": idx,
        }
        out.append(row)
    return out


def main() -> int:
    ap = argparse.ArgumentParser(description="Python AST function-catalog extractor.")
    ap.add_argument("--root", required=True, help="Root of the codebase to scan.")
    ap.add_argument("--shared", help="Secondary package root tagged as package='shared'.")
    ap.add_argument("--touched", help="Path to a JSON array of touched-in-window files.")
    ap.add_argument("--output", help="Write JSON to this path. Default: stdout.")
    ap.add_argument(
        "--min-body-lines", type=int, default=DEFAULT_MIN_BODY_LINES,
        help=(
            "Functions whose normalized body has fewer than this many lines emit "
            "body_hash / body_lines / body_line_count / body_length as null. "
            "Default: 3."
        ),
    )
    args = ap.parse_args()

    root = Path(args.root).resolve()
    shared = Path(args.shared).resolve() if args.shared else None
    touched: set[str] = set()
    if args.touched:
        with open(args.touched, encoding="utf-8") as fh:
            touched = set(json.load(fh))

    extractor_dir = Path(__file__).resolve().parent
    source_sha = compute_source_sha(extractor_dir)
    generated_at = datetime.datetime.now(datetime.timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")

    main_files = list(walk_python_files(root))
    sys.stderr.write(f"main: {len(main_files)} files\n")
    entries: list[dict] = []
    errors = 0
    for f in main_files:
        try:
            entries.extend(extract_from_file(f, "main", root, touched, args.min_body_lines))
        except Exception as exc:
            errors += 1
            sys.stderr.write(f"  ERR {f}: {exc}\n")

    if shared:
        shared_files = list(walk_python_files(shared))
        sys.stderr.write(f"shared: {len(shared_files)} files\n")
        for f in shared_files:
            try:
                entries.extend(extract_from_file(f, "shared", shared, set(), args.min_body_lines))
            except Exception as exc:
                errors += 1
                sys.stderr.write(f"  ERR {f}: {exc}\n")

    entries.sort(key=lambda r: (r["package"], r["file"], r["line"], r["name"]))

    payload = {
        "schema_version": SCHEMA_VERSION,
        "extractor": make_extractor_block("function-catalog", source_sha),
        "fingerprint_v": FINGERPRINT_V,
        "generated_at": generated_at,
        "entries": entries,
    }

    sys.stderr.write(f"\nTotal entries: {len(entries)} (errors: {errors})\n")
    write_catalog(args.output, payload)
    return 0 if len(main_files) > 0 else 1


if __name__ == "__main__":
    sys.exit(main())
